src/main/java/formflow/library/data/SubmissionRepositoryService.java

```diff
@@ -17,67 +19,137 @@
-    return submission.map(value -> encryptionService.decrypt(value));
-  }
-
-  /**
-   * Removes the CSRF from the Submission's input data, if found.
-   *
-   * @param submission submission to remove the CSRF from, not null
-   */
-  public void removeFlowCSRF(Submission submission) {
-    submission.getInputData().remove("_csrf");
-  }
-
-  /**
-   * Removes the CSRF from a particular Submission's Subflow's iteration data, if found.
-   * <p>
-   * This will remove the CSRF from all the iterations in the subflow.
-   * </p>
-   *
-   * @param submission  submission to look for subflows in, not null
-   * @param subflowName the subflow to remove the CSRF from, not null
-   */
-  public void removeSubflowCSRF(Submission submission, String subflowName) {
-    var subflowArr = (ArrayList<Map<String, Object>>) submission.getInputData().get(subflowName);
-
-    if (subflowArr != null) {
-      for (var entry : subflowArr) {
-        entry.remove("_csrf");
-      }
+
+    /**
+     * Removes the CSRF from a particular Submission's Subflow's iteration data, if found.
+     * <p>
+     * This will remove the CSRF from all the iterations in the subflow.
+     * </p>
+     *
+     * @param submission  submission to look for subflows in, not null
+     * @param subflowName the subflow to remove the CSRF from, not null
+     */
+    public void removeSubflowCSRF(Submission submission, String subflowName) {
+        var subflowArr = (ArrayList<Map<String, Object>>) submission.getInputData().get(subflowName);
+
+        if (subflowArr != null) {
+            for (var entry : subflowArr) {
+                entry.remove("_csrf");
+            }
+        }
+    }
+
+    /**
+     * generateAndSetUniqueShortCode generates a read-only unique code for the submission. The short code generation is
+     * configurable via {@link formflow.library.config.submission.ShortCodeConfig}:
+     * <p>
+     * length (default = 6)
+     * <p>
+     * characterset (alphanumeric, numeric, alpha | default = alphanumeric)
+     * <p>
+     * forced uppercasing (true, false | default = true)
+     * <p>
+     * creation point in {@link formflow.library.config.submission.ShortCodeConfig} (creation, submission | default = submission)
+     * <p>
+     * prefix (default = null)
+     * <p>
+     * suffix (default = null)
+     * <p>
+     * This method will check if the generated code exists in the database, and keep trying to create a unique code, before
+     * persisting and returning the newly generated code-- therefore it is very important to ensure the configuration allows for a
+     * suitably large set of possible codes for the application.
+     *
+     * @param submission the {@link formflow.library.ScreenController} for which the short code will be generated and saved
+     */
+    public synchronized void generateAndSetUniqueShortCode(Submission submission) {
+
+        if (submission.getShortCode() != null) {
+            return;
```

Comment:
We should log here as ideally we shouldn't bit this return, but it would be to know if we are.